hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/GrpcContainerUploader.java

```diff
@@ -53,18 +55,32 @@ public GrpcContainerUploader(
   public OutputStream startUpload(long containerId, DatanodeDetails target,
       CompletableFuture<Void> callback, CopyContainerCompression compression)
       throws IOException {
-    GrpcReplicationClient client =
-        new GrpcReplicationClient(target.getIpAddress(),
-            target.getPort(Port.Name.REPLICATION).getValue(),
-            securityConfig, certClient, compression);
-    StreamObserver<SendContainerRequest> requestStream = client.upload(
-        new SendContainerResponseStreamObserver(containerId, target, callback));
-    return new SendContainerOutputStream(requestStream, containerId,
-        GrpcReplicationService.BUFFER_SIZE, compression);
+    GrpcReplicationClient client = null;
```

Comment:
Should the code around creating `client` and uploading be refactored to be done in one place, e.g. in `PushReplicator.replicate`.

```
try (GrpcReplicationClient client  = grpcContainerUploader.createClient()) {
  try (OutputStream output : client.upload(....)) {
     source.copyData(containerID, output, compression);
     ...
  }
}
```
That way we wouldn't need to care about passing `client` around and closing it in multiple places.

That's just a suggestion to simplify the model. Looks like in this PR you already take good care of closing resources.

Reply:
Thanks for the idea.  I prefer to keep `PushReplicator` independent of the gRPC implementation, similar to `DownloadAndImportReplicator` (pull replication).  But I agree that passing around `client` just to close it is a hack, created HDDS-8119 to get rid of it.